### PR TITLE
[Bug]: backend search update fails in high concurrency situations

### LINKED ISSUE #13507: [Bug]: backend search update fails in high concurrency situations
### Expected behavior

Search backend updates work.

### Actual behavior

We're seeing hundreds of deadlock messages and our workers constantly crashing.

### Steps to reproduce

Have a high rate of update (say, import 30k objects) and multiple workers (we use 4). The rate of deadlocks is very high which ends up crashing our workers (which rely on DBAL based messenger).

Looking at the code which saves, I can see it's doing a lot of weird work, catching deadlock and then trying to do it at random (?).

The issue is the query doing it is doing
```sql
INSERT ... ON DUPLICATE KEY UPDATE
```

which is [acquiring different types of locks](https://medium.com/@mingwho/mysql-deadlock-caused-by-insert-on-duplicate-key-update-893ef30993a8) than just a plain `INSERT`.

The solution might be to do it yourself by trying an INSERT and, if it fails, do an UPDATE. This wouldn't acquire [a gap lock](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking.html#innodb-gap-locks) and should reduce the number of deadlocks significantly, probably making all the current deadlock-avoidance related code in the search DAO obsolete.

### PR DESCRIPTION
**Pull request by mcop1**
Resolves #13507

`insertOrUpdate` won´t be removed for now, see following tickets: 
https://github.com/pimcore/pimcore/issues/14446
https://github.com/pimcore/pimcore/issues/14448

### CODE REVIEW

File: models/DataObject/Classificationstore/KeyGroupRelation/Dao.php [hunk 1]
```diff
@@ -96,6 +96,6 @@ public function update(): void
             }
         }
 
-        Helper::insertOrUpdate($this->db, self::TABLE_NAME_RELATIONS, $data);
+        Helper::upsert($this->db, self::TABLE_NAME_RELATIONS, $data, ['groupId', 'keyId']);
```
Comment (fashxp) [inline models/DataObject/Classificationstore/KeyGroupRelation/Dao.php]: why not use `$this->getPrimaryKey()` here?

File: models/DataObject/Concrete/Dao.php [hunk 2]
```diff
@@ -278,12 +278,8 @@ public function update(bool $isUpdate = null): void
...
             }
         }
-
-        if ($isUpdate) {
-            Helper::insertOrUpdate($this->db, 'object_store_' . $this->model->getClassId(), $data);
-        } else {
-            $this->db->insert('object_store_' . $this->model->getClassId(), Helper::quoteDataIdentifiers($this->db, $data));
-        }
+        $tableName = 'object_store_' . $this->model->getClassId();
+        Helper::upsert($this->db, $tableName, $data, $this->getPrimaryKey($tableName));
```
Comment (fashxp) [inline models/DataObject/Concrete/Dao.php]: Isn't that a change of behavior? 
Previously, there was an exception thrown, if `$isUpdate === false` and ID was already in DB. Now, it silently updates the existing record.

File: models/DataObject/Localizedfield/Dao.php [hunk 3]
```diff
@@ -163,11 +163,9 @@ public function save(array $params = []): void
                 'language' => $language,
             ];
 
-            if ($container instanceof DataObject\Objectbrick\Definition) {
+            if ($container instanceof DataObject\Objectbrick\Definition || $container instanceof DataObject\Fieldcollection\Definition) {
                 $insertData['fieldname'] = $context['fieldname'];
-            } elseif ($container instanceof DataObject\Fieldcollection\Definition) {
-                $insertData['fieldname'] = $context['fieldname'];
-                $insertData['index'] = $context['index'];
+                $insertData['index'] = $context['index'] ?? 0;
```
Comment (fashxp) [inline models/DataObject/Localizedfield/Dao.php:168]: but according brick tables do not have the index column, right? 
so, there is actually no sense in setting that value?

**Review by fashxp — CHANGES_REQUESTED** [review]

File: models/DataObject/Concrete/Dao.php (hunk 2 above)
Comment (mcop1) [reply models/DataObject/Concrete/Dao.php]: True, if `$isUpdate` is set to false and the key is already existing in the table it wouldn´t throw an exception anymore. But then the question would be how that can happen. 

Anyway, I reverted the change and substituted the `Helper::insertOrUpdate` call only.

File: models/DataObject/Classificationstore/KeyGroupRelation/Dao.php (hunk 1 above)
Comment (mcop1) [reply models/DataObject/Classificationstore/KeyGroupRelation/Dao.php]: Missed that one after changing the approach.

File: models/DataObject/Localizedfield/Dao.php (hunk 3 above)
Comment (mcop1) [reply models/DataObject/Localizedfield/Dao.php:168]: The column is there, but not used, but still part of the primary key. 
It´s the same code for object bricks and field collections:
https://github.com/pimcore/pimcore/blob/f3275ffd2352e437a67df8bf7edaa2f675cd3831/models/DataObject/Localizedfield/Dao.php#L801

Comment (fashxp) [reply models/DataObject/Localizedfield/Dao.php:168]: true...

**Review by fashxp — APPROVED** [review]

**Pull request merged**